docker/Dockerfile

```diff
@@ -0,0 +1,12 @@
+# syntax=docker/dockerfile:1
+FROM golang:1.21.3-bullseye as builder
+WORKDIR /src
+
+COPY go.mod .
+COPY *.go .
+ARG CGO_ENABLED=0
```

Comment:
[optional nit]: i think it's nice to list ARGs basically at the start of the file - just below the FROM here since we're using it in the build steps.

Reply:
Yeah I was a bit torn between either putting at the top of the file, or closer to the place that it was affecting. Or I could have inlined it with something like

```
RUN CGO_ENABLED=0 go build -o /bin/sms-provider-stub .
```

but this also felt less readable.

Anyway, back to the ARGs:
Is putting ARGs at the top a pattern that's recommended/followed either generally or in our Dockerfiles?

Reply:
Looking across a few of our dockerfiles, ARG is _generally_ grouped at the top of the file, near FROM or ENV steps.

<img width="788" alt="image" src="https://github.com/alphagov/notifications-sms-provider-stub/assets/2920760/a2cf035a-e570-4125-9fcd-9947bc175c3a">

---

<img width="797" alt="image" src="https://github.com/alphagov/notifications-sms-provider-stub/assets/2920760/91a5c2f6-ac75-4d1a-b0d1-4920cf76f82b">

---

<img width="739" alt="image" src="https://github.com/alphagov/notifications-sms-provider-stub/assets/2920760/b3ef3d31-e5ce-4a30-b764-0d1f852b2df0">

---

<img width="788" alt="image" src="https://github.com/alphagov/notifications-sms-provider-stub/assets/2920760/c2556fe1-5d31-4c24-b6a2-7e347822aef1">

Reply:
Oh nice, thanks for doing the digging :) 

However your last point makes me insist on leaving it there (emphasis mine):

> but as an (**overrideable**) input to the build in my head it makes sense for it to be near the top for discoverability.

It is not supposed to be overrideable, it is necessary for the build to work. I'm just using `ARG` because it is only relevant for the build phase and we don't want it to be inherited by or persisted into subsequent images, and because the inline alternative is uglier.

I will add a comment though to make it explicit that it is required.

Reply:
Amended the commit to add the comment.